src/main/java/com/practice/slideshow/service/ImageService.java

```diff
@@ -28,42 +40,72 @@ public class ImageService {
    * @return The created ImageEntity.
    */
   @Transactional
-  public ImageEntity addImage(String url, int duration) {
+  public ImageResponse addImage(String url, int duration) {
     urlValidationService.validateImageUrl(url);
+    log.info("Adding new image with URL: {} and duration: {}", url, duration);
 
-    ImageEntity image = ImageEntity.builder()
+    ImageEntity savedImage = imageRepository.save(ImageEntity.builder()
         .url(url)
         .duration(duration)
-        .build();
+        .build());
 
-    ImageEntity savedImage = imageRepository.save(image);
-    kafkaLoggingService.logAction("ADD_IMAGE",
-        String.format("Image ID: %d, URL: %s", savedImage.getId(), url));
+    kafkaLoggingService.logAction(new LogEvent(savedImage.getId(), LogEventType.ADD_IMAGE));
+    log.info("Image added successfully with ID: {}", savedImage.getId());
 
-    return savedImage;
+    return ImageMapper.mapFromEntity(savedImage);
   }
 
   /**
    * Deletes an image by its ID.
    *
    * @param id The image ID.
    */
-  @Transactional
   public void deleteImage(Long id) {
-    if (!imageRepository.existsById(id)) {
-      throw new ResourceNotFoundException("Image not found, id: " + id);
-    }
-    imageRepository.deleteById(id);
-    kafkaLoggingService.logAction("DELETE_IMAGE", String.format("Deleted Image ID: %d", id));
+    log.info("Attempting to delete image with ID: {}", id);
+    ImageEntity image = imageRepository.findById(id)
+        .orElseThrow(() -> {
+          log.error("Image with ID: {} not found.", id);
+          return new ResourceNotFoundException("Image not found, id: " + id);
+        });
+    kafkaLoggingService.logAction(new LogEvent(image.getId(), LogEventType.DELETE_IMAGE));
+    imageRepository.delete(image);
+    log.info("Image deleted successfully for ID: {}", id);
   }
 
   /**
    * Searches images by a keyword, matching the URL or duration.
-   * @param keyword search keyword
-   * @return list of matching images
+   *
+   * @param keyword The search keyword.
+   * @return A list of matching images.
    */
   @Transactional(readOnly = true)
   public List<ImageEntity> searchImages(String keyword) {
+    log.info("Searching for images with keyword: {}", keyword);
     return imageRepository.findAll(ImageSpecification.hasUrlContaining(keyword));
   }
+
+  /**
+   * Searches for images by a keyword and maps results to a response format.
+   *
+   * @param keyword The search keyword.
+   * @return A response containing the list of matching images.
+   */
+  @Transactional(readOnly = true)
+  public ImageSearchResponse searchImagesWithResults(String keyword) {
+    log.info("Searching for images and mapping results with keyword: {}", keyword);
+    var images = searchImages(keyword);
+    var results = images.stream().map(i -> {
+      var associatedSlideshows = slideshowRepository.findByImageId(i.getId())
```

Comment:
it does not look right to have separate call for each image to find associated image could you please create separate query to retrieve all related data